src/raw/mod.rs

```diff
@@ -2141,9 +2208,28 @@ impl<A: Allocator + Clone> RawTableInner<A> {
         self.bucket_mask == 0
     }
 
+    /// Attempts to allocate a new hash table with at least enough capacity
+    /// for inserting the given number of elements without reallocating,
+    /// and return it inside ScopeGuard to protect against panic in the hash
+    /// function.
+    ///
+    /// # Note
+    ///
+    /// It is recommended (but not required):
+    ///
+    /// * That the new table's `capacity` be greater than or equal to `self.items`.
+    ///
+    /// * The `table_layout` is the same [`TableLayout`] as the `TableLayout` that
+    ///   was used to allocate this table.
+    ///
+    /// If `table_layout` does not match the `TableLayout` that was used to allocate
+    /// this table, then using `mem::swap` with the `self` and the new table returned
+    /// by this function results in [`undefined behavior`].
+    ///
+    /// [`undefined behavior`]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
     #[allow(clippy::mut_mut)]
     #[inline]
-    unsafe fn prepare_resize(
+    fn prepare_resize(
```

Comment:
It occurs to me that a lot of these methods that mutate data on the heap should probably take `&mut self`, otherwise there is an additional safety requirement that no other thread is concurrently accessing the data.

Reply:
I can try to go through all the functions (after the PR), although it is probably won’t be able to do this with all of them.

Specifically, this function can be left as it is (with `&self`), since it does not change the original table (neither before nor after this pull request).